kani-compiler/src/kani_middle/transform/check_aliasing/instrumentation.rs

```diff
@@ -0,0 +1,446 @@
+        let fn_pointers = &mut self.fn_pointers;
+        let body = &mut self.body;
+        let span = self.span.clone();
+        let instance = cache.register(tcx, callee)?;
+        let func_local = fn_pointers
+            .entry(*instance)
+            .or_insert_with(|| body.new_local(instance.ty(), span, Mutability::Not));
+        Ok(*func_local)
+    }
+
+    /// Call at source and insert position, using the arguments
+    /// in args and returning into "dest".
+    /// This differs from Mutable Body's call in that the
+    /// function name is cached.
+    pub fn call(&mut self, callee: Signature, args: Vec<Local>, dest: Local) -> Result<()> {
+        let func_local = self.register_fn(callee)?;
+        let new_bb = self.body.blocks().len();
+        let span = self.body.blocks()[self.min_processed.bb()].terminator.span;
+        let callee_op = Operand::Copy(Place::from(func_local));
+        let args = args
+            .into_iter()
+            .map(|v| Operand::Copy(Place { local: v, projection: vec![] }))
+            .collect();
+        let destination = Place::from(dest);
+        let kind = TerminatorKind::Call {
+            func: callee_op,
+            args,
+            destination,
+            target: Some(new_bb),
+            unwind: UnwindAction::Terminate,
+        };
+        let terminator = Terminator { kind, span };
+        let source = &mut self.ghost_index;
+        self.body.insert_terminator(source, InsertPosition::After, terminator);
+        Ok(())
+    }
+
+    /// Instrument an assignment to a local
+    pub fn assign_pointer(&mut self, lvalue: Local, rvalue: Local) {
+        let source = &mut self.ghost_index;
+        let position = InsertPosition::After;
+        self.body.assign_to(
+            Place::from(lvalue),
+            Rvalue::AddressOf(Mutability::Not, Place::from(rvalue)),
+            source,
+            position,
+        );
+    }
+
+    /// For some local, say let x: T;
+    /// instrument it with the functions that initialize the stack:
+    /// let ptr_x: *const T = &raw const x;
+    /// initialize_local(ptr_x);
+    pub fn instrument_local(&mut self, local: Local) -> Result<()> {
+        let ty = self.body.locals()[local].ty;
+        let ptr_ty = Ty::new_ptr(ty, Mutability::Not);
+        let span = self.span.clone();
+        let body = &mut self.body;
+        let local_ptr = self
+            .meta_stack
+            .entry(local)
+            .or_insert_with(|| body.new_local(ptr_ty, span, Mutability::Not));
+        let local_ptr = *local_ptr;
+        self.assign_pointer(local_ptr, local);
+        self.call(
+            Signature::new("KaniInitializeLocal", &[GenericArgKind::Type(ty)]),
+            vec![local_ptr],
+            self.unit,
+        )?;
+        Ok(())
+    }
+
+    /// Split at the minimum processed instruction,
+    /// allowing instrumentation of ghost code following
+    /// that source instruction.
+    pub fn process_instruction(&mut self) {
```

Comment:
I believe this function could use some of the functionality of `insert_bb` from `MutableBody`.

Reply:
This is more general than insert_bb, even in just the statement case -- insert_bb grabs mut target ref, which  may only be one with a target. This is instead just splitting the basic block -- not inserting a basic block after another basic block